controllers/metering/controllers/metering_controller.go

```diff
@@ -20,210 +20,238 @@ import (
-	if errors.IsNotFound(err) {
-		r.Logger.Info("creat metering")
-		metering = meteringv1.Metering{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      fmt.Sprintf("metering-%v", ns.Name),
-				Namespace: SealosSystemNamespace,
-			},
-			Spec: meteringv1.MeteringSpec{
-				Namespace: ns.Name,
-				Owner:     owner,
-				Resources: DefaultResourceQuota(),
-			},
-			Status: meteringv1.MeteringStatus{
-				TotalAmount:      0,
-				LatestUpdateTime: time.Now().Unix(),
-			},
+	if time.Now().Unix()-metering.Status.LatestUpdateTime >= int64(time.Minute.Seconds())*int64(metering.Spec.TimeInterval) {
+		r.Logger.Info("enter update metering", "metering name:", req.Name, "metering namespace:", req.Namespace, "lastUpdate Time", metering.Status.LatestUpdateTime, "now", time.Now().Unix(), "diff", time.Now().Unix()-metering.Status.LatestUpdateTime, "interval", int64(time.Minute.Seconds())*int64(metering.Spec.TimeInterval))
+		totalAccount, err := r.CalculateCost(ctx, &metering)
+		if err != nil {
+			r.Logger.Error(err, err.Error())
+			return ctrl.Result{}, err
 		}
-		if err := r.Create(ctx, &metering); err != nil {
-			return meteringv1.Metering{}, fmt.Errorf("create metering failed: %v", err)
+
+		if err := r.syncAccountBalance(ctx, metering.Spec.Owner, totalAccount, metering.Status.SeqID); err != nil {
+			r.Logger.Error(err, err.Error())
+			return ctrl.Result{}, fmt.Errorf("meteringName:%v,amount:%v,err:%v", metering.Name, totalAccount, err)
+		}
+		if err := r.updateBillingList(ctx, totalAccount, &metering); err != nil {
+			r.Logger.Error(err, err.Error())
+			return ctrl.Result{}, err
 		}
-	} else if err != nil {
-		r.Logger.Error(err, "Failed to get Metering")
-		return meteringv1.Metering{}, client.IgnoreNotFound(err)
 	}
-	return metering, nil
+
+	return ctrl.Result{Requeue: true, RequeueAfter: time.Minute * time.Duration(r.MeteringInterval)}, nil
 }
 
-func (r *MeteringReconcile) syncResourceQuota(ctx context.Context, metering meteringv1.Metering) error {
-	quota := &corev1.ResourceQuota{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      metering.Name,
-			Namespace: metering.Spec.Namespace,
-		},
+func (r *MeteringReconcile) CalculateCost(ctx context.Context, metering *meteringv1.Metering) (int64, error) {
+	var resourceMsgs []meteringv1.ResourceMsg
+	infraAmount, err := r.calculateInfraCost(ctx, metering)
+	if err != nil {
+		r.Logger.Error(err, "Failed to get InfraCost")
+	} else {
+		resourceMsgs = append(resourceMsgs, meteringv1.ResourceMsg{
+			ResourceName: "infra",
+			Amount:       float64(infraAmount),
+		})
 	}
-	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, quota, func() error {
-		quota.Spec.Hard = metering.Spec.Resources
+
+	var totalAmount float64
+	for resourceName, resourceValue := range metering.Spec.Resources {
+		if _, ok := metering.Spec.Resources[resourceName]; ok {
+			amount := float64(resourceValue.Used.MilliValue()) * float64(resourceValue.Price) / float64(resourceValue.Unit.MilliValue())
```

Comment:
maybe MilliValue is use less?

Reply:
like cpu，if use value it can not expressed as 800mi